csharp/src/Ice/IncomingRequestFrame.cs

```diff
@@ -56,7 +56,19 @@ public IncomingRequestFrame(Communicator communicator, ArraySegment<byte> data)
             IsIdempotent = istr.ReadOperationMode() != OperationMode.Normal;
             Context = istr.ReadContext();
             Payload = Data.Slice(istr.Pos);
-            Encoding = istr.StartEncapsulation();
+            (Encoding encoding, int size) = istr.ReadEncapsulationHeader();
+            if (size != Payload.Count)
+            {
+                throw new InvalidDataException($"invalid encapsulation size: {size}");
+            }
+            Encoding = encoding;
```

Comment:
For another PR, but we should consider slicing Data here to match the end of the encapsulation. This would allow to support batch requests.